### PR TITLE
feat: Implement the Buf to avoid extra memory allocation

### LINKED ISSUE #4065: Implement the Buf to avoid extra memory allocation
We need something like this: https://github.com/apache/opendal/blob/dd70b3c706f1689c2fe9455405cbd1e773c483ab/core/src/types/buffer.rs#L114-L122

_Originally posted by @WenyXu in https://github.com/GreptimeTeam/greptimedb/pull/4038#discussion_r1618113196_

### PR DESCRIPTION
**Pull request by ozewr**
I hereby agree to the terms of the [GreptimeDB CLA](https://github.com/GreptimeTeam/.github/blob/main/CLA.md).

## Refer to a related PR or issue link (optional)

close #4065 

## What's changed and what's your intention?
Impl new function `into_buffer` for Entry in `/home/wr/workspace/greptimedb/src/store-api/src/logstore/entry.rs`.
This function returns a `buffer` that comes from the `opendal` library, but is reused in the `object-store`.
Use `into_buffer` to avoid extra memory allocation in file `src/mito2/src/wal/entry_reader.rs`.

## Checklist

- [ ] I have written the necessary rustdoc comments.
- [ ] I have added the necessary unit tests and integration tests.
- [ ] This PR requires documentation updates.

### CODE REVIEW

File: src/object-store/src/lib.rs [hunk 1]
```diff
@@ -24,3 +24,4 @@ pub mod manager;
 mod metrics;
 pub mod test_util;
 pub mod util;
+pub use opendal;
```
Comment (waynexia) [inline src/object-store/src/lib.rs:27]: nit:
```suggestion
pub use opendal;

```

File: src/mito2/src/wal/entry_reader.rs [hunk 2]
```diff
@@ -29,9 +29,8 @@ pub(crate) fn decode_raw_entry(raw_entry: Entry) -> Result<(EntryId, WalEntry)>
     let region_id = raw_entry.region_id();
     ensure!(raw_entry.is_complete(), CorruptedEntrySnafu { region_id });
     // TODO(weny): implement the [Buf] for return value, avoid extra memory allocation.
```
Comment (waynexia) [inline src/mito2/src/wal/entry_reader.rs:31]: We can remove this todo now cc @WenyXu 
```suggestion
```

**Review by waynexia — APPROVED** [review]
Looks good to me, thanks @ozewr !

**Review by waynexia — APPROVED** [review]
Looks good to me, thanks @ozewr !

File: src/store-api/src/logstore/entry.rs [hunk 3]
```diff
@@ -143,6 +144,15 @@ impl Entry {
         }
     }
 
+    pub fn into_buffer(self) -> Buffer {
+        match self {
+            Entry::Naive(entry) => Buffer::from(entry.data),
+            Entry::MultiplePart(entry) => Buffer::from(bytes::Bytes::from_iter(entry.parts.into_iter().flatten())),
+
+        }
```
Comment (WenyXu) [inline src/store-api/src/logstore/entry.rs:152]:
```suggestion
        match self {
            Entry::Naive(entry) => Buffer::from(entry.data),
            Entry::MultiplePart(entry) => {
                Buffer::from_iter(entry.parts.into_iter().map(bytes::Bytes::from))
            }
        }
```

File: src/store-api/src/logstore/entry.rs [hunk 4]
```diff
@@ -143,6 +144,15 @@ impl Entry {
         }
     }
 
+    pub fn into_buffer(self) -> Buffer {
```
Comment (WenyXu) [inline src/store-api/src/logstore/entry.rs:147]: I prefer moving function into `mito2` crate.

Comment (ozewr) [reply src/store-api/src/logstore/entry.rs:148]: But the Entry struct in the object-store crate. Wouldn’t it be more intuitive to implement it there?

Comment (ozewr) [reply src/store-api/src/logstore/entry.rs:148]: But the `Entry` struct is in the `store-api` crate. Wouldn’t it be more intuitive to implement it there?

Comment (WenyXu) [reply src/store-api/src/logstore/entry.rs:148]: I think it's better to keep `store-api` crate simple.t

Comment (ozewr) [reply src/store-api/src/logstore/entry.rs:148]: ok I will write it in `src/mito2/src/wal/entry_reader.rs` , by the way how to fix  Toml check ,I aready run `make fmt-toml`,but it still error in CI. I also run `taplo format check ` on my pc,but no Error.
```shell
taplo format --check
 INFO taplo:format_files:load_config: found configuration file path="/home/wr/workspace/greptimedb/taplo.toml"
 INFO taplo:format_files:collect_files: found files total=61 excluded=0 cwd="/home/wr/workspace/greptimedb"
```

File: src/object-store/src/lib.rs (hunk 1 above)
Comment (WenyXu) [inline src/object-store/src/lib.rs:27]:
```suggestion
pub use opendal::Buffer;
```

**Review by WenyXu — APPROVED** [review]
Rest LGTM

File: src/object-store/src/lib.rs [hunk 5]
```diff
@@ -24,3 +24,4 @@ pub mod manager;
 mod metrics;
 pub mod test_util;
 pub mod util;
+pub use opendal::Buffer;
```
Comment (WenyXu) [inline src/object-store/src/lib.rs:27]: nit: we can move it to L16, like 
```rust 
pub use opendal::{
    services, Buffer, Builder as ObjectStoreBuilder, Entry, EntryMode, Error, ErrorKind,
    FuturesAsyncReader, FuturesAsyncWriter, Lister, Metakey, Operator as ObjectStore, Reader,
    Result, Writer,
};
```

**Pull request merged**